README.md
# eslint-config-yoctol

[![NPM version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Greenkeeper badge](https://badges.greenkeeper.io/Yoctol/eslint-config-yoctol.svg)](https://greenkeeper.io/)
[![Dependency Status][david_img]][david_site]

> Yoctol Info. ESLint config, fork from [eslint-config-airbnb](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb)

## Usage

### eslint-config-yoctol

Our default export contains all of our ESLint rules, including EcmaScript 6+
and React. It requires `eslint`, `eslint-plugin-import`, `eslint-plugin-react`, `eslint-plugin-react-hooks`, `eslint-plugin-jsx-a11y`, `eslint-plugin-prettier`, `eslint-plugin-sort-imports-es6-autofix` and `prettier`.

1. Install the correct versions of each package, which are listed by the command:

```sh
npm info "eslint-config-yoctol@latest" peerDependencies
```

Linux/OSX users can simply run

```sh
(
  export PKG=eslint-config-yoctol;
  npm info "$PKG@latest" peerDependencies --json | command sed 's/[\{\},]//g ; s/: /@/g' | xargs npm install --save-dev "$PKG@latest"
)
```

Which produces and runs a command like:

```sh
npm install --save-dev eslint-config-yoctol eslint@^#.#.# eslint-plugin-jsx-a11y@^#.#.# eslint-plugin-import@^#.#.# eslint-plugin-react@^#.#.# eslint-plugin-react-hooks@^#.#.# eslint-plugin-prettier@^#.#.# eslint-plugin-sort-imports-es6-autofix@^#.#.#
```

Windows users can either install all the peer dependencies manually, or use the [install-peerdeps](https://github.com/nathanhleung/install-peerdeps) cli tool.

```sh
npm install -g install-peerdeps
install-peerdeps --dev eslint-config-yoctol
```

The cli will produce and run a command like:

```sh
npm install --save-dev eslint-config-yoctol eslint@^#.#.# eslint-plugin-jsx-a11y@^#.#.# eslint-plugin-import@^#.#.# eslint-plugin-react@^#.#.# eslint-plugin-react-hooks@^#.#.# eslint-plugin-prettier@^#.#.# eslint-plugin-sort-imports-es6-autofix@^#.#.#
```

2. Add `"extends": "yoctol"` to your .eslintrc

### eslint-config-yoctol/base

This entry point is deprecated. See [eslint-config-yoctol-base](https://npmjs.com/eslint-config-yoctol-base).

### eslint-config-yoctol/legacy

This entry point is deprecated. See [eslint-config-yoctol-base](https://npmjs.com/eslint-config-yoctol-base).

See [Airbnb's Javascript styleguide](https://github.com/yoctol/javascript) and
the [ESlint config docs](http://eslint.org/docs/user-guide/configuring#extending-configuration-files)
for more information.

## Rules be overwritten

### react

```
- react/destructuring-assignment: 'off'
- react/sort-prop-types: ['error', {
  callbacksLast: true,
  ignoreCase: false,
  requiredFirst: false,
}]
- react/no-deprecated: ['warn']
- react/prefer-stateless-function: 'off'
- react/jsx-filename-extension: ['error', { extensions: ['.js'] }]
- react-hooks/rules-of-hooks: 'error'
- react-hooks/exhaustive-deps: 'error'
```

[npm-image]: https://badge.fury.io/js/eslint-config-yoctol.svg
[npm-url]: https://npmjs.org/package/eslint-config-yoctol
[travis-image]: https://travis-ci.org/Yoctol/eslint-config-yoctol.svg
[travis-url]: https://travis-ci.org/Yoctol/eslint-config-yoctol
[david_img]: https://david-dm.org/Yoctol/eslint-config-yoctol.svg
[david_site]: https://david-dm.org/Yoctol/eslint-config-yoctol


CHANGELOG.md
# CHANGELOG

## 0.26.3 / 2021-09-17

- [deps] support `eslint-plugin-prettier` >= 4.

## 0.26.2 / 2021-03-31

- [fix] turn off react rules that conflict with prettier

## 0.26.1 / 2021-03-30

- [deps] update `eslint-config-yoctol-base` to `0.24.1`
- [deps] `eslint-plugin-sort-imports-es6-autofix^0.6.0` as peerDeps

## 0.26.0 / 2021-03-29

- [deps] update `eslint-config-yoctol-base` to `0.24.0`
- [deps] use `eslint>=7.7.0` and `eslint-plugin-react-hooks^4.0.0` as peerDeps
- [deps] bump other deps

## 0.25.1 / 2020-06-20

- [deps] support ESLint v7

## 0.25.0 / 2020-04-28

- [deps] update `eslint-config-yoctol-base` to `0.23.0`
- [deps] move `prettier` to peerDeps
- [deps] use `prettier>=2` as peerDeps
- [deps] bump other deps

## 0.24.0 / 2020-02-04

- [breaking] requires `node>=10`
- [deps] update `eslint-config-yoctol-base` to `0.22.0`
- [deps] bump other deps

## 0.23.3 / 2019-09-05

- [fix] update eslint-config-yoctol-base to support eslint-plugin-sort-imports-es6-autofix 0.5.0

## 0.23.2 / 2019-09-05

- [deps] update dependency eslint-plugin-sort-imports-es6-autofix to 0.5.0
- [deps] bump other deps

## 0.23.1 / 2019-08-27

- chore: bump deps for security alert (#116)

## 0.23.0 / 2019-08-13

- [new] add `eslint` v6 support
- [deps] bump other deps

## 0.22.1 / 2019-05-30

- [deps] bump deps

## 0.22.0 / 2019-05-30

- [CI] remove node 6 and add node 10
- [deps] bump deps

## 0.21.1 / 2019-02-22

- [new] enabled `react-hooks/exhaustive-deps`
- [new] set `react/destructuring-assignment` to `off`
- [fix] use yarn upgrade to upgrade vulnerable lodash version
- [deps] update dependencies

## 0.21.0 / 2019-01-29

- [deps] update `eslint-config-yoctol-base` to `0.19.0`
- [deps] update `prettier` to `1.16.1`
- [deps] peerDependencies `eslint` to `v5.12.1`
- [deps] peerDependencies `eslint-plugin-jsx-a11y` to `v6.2.0`
- [deps] peerDependencies `eslint-plugin-import` to `v2.15.0`
- [deps] peerDependencies `eslint-plugin-prettier` to `v3.0.1`
- [deps] peerDependencies `eslint-plugin-react` to `v7.12.4`

## 0.20.0 / 2018-11-13

- [new] add `eslint-plugin-react-hooks` rules
- [deps] update `eslint` to `v5.9.0`

## 0.19.0 / 2018-10-24

- [deps] peerDependencies `eslint-plugin-prettier` to `v3.0.0`

## 0.18.2 / 2018-09-16

- [deps] update `eslint-config-yoctol-base` to `0.17.2`

## 0.18.1 / 2018-08-21

- [new] add `eslint` v5 support
- [deps] update `eslint-config-yoctol-base` to `0.17.1`
- [deps] update `eslint-config-airbnb` to `17.1.0`
- [deps] update `eslint-plugin-import` to `2.14.0`
- [deps] update `eslint-plugin-react` to `7.11.1`
- [deps] update `prettier` to `1.14.2`

## 0.18.0 / 2018-07-23

- [deps] update `eslint-config-yoctol-base` to `v0.17.0`

## 0.17.1 / 2018-06-26

- [fix] overwrite rules with eslint-config-prettier/react

## 0.17.0 / 2018-06-25

- [deps] update `eslint-config-yoctol-base` to `v0.16.0`
- [deps] update `eslint-config-airbnb` to `v17.0.0`

## 0.16.1 / 2018-02-12

- [deps] update `eslint-config-yoctol-base` to `v0.15.1`

## 0.16.0 / 2018-01-28

- [deps] update `eslint-config-yoctol-base` to `v0.15.0`
- [deps] update `eslint` and eslint plugins

## 0.15.1 / 2017-10-20

- [deps] update `eslint-config-airbnb` to `v16.1.0`
- [deps] update `eslint-plugin-import` to `v2.8.0`
- [deps] update `eslint-plugin-jsx-a11y` to `v6.0.2`
- [deps] update `eslint` to `v4.9.0`

## 0.15.0 / 2017-10-15

- [deps] update `eslint-config-airbnb` to `v16.0.0`
- [deps] update `eslint-config-yoctol-base` to `v0.14.1`
- [deps] update `eslint-plugin-react` to `v7.4.0`
- [deps] update `eslint` to `v4.8.0`

## 0.14.0 / 2017-08-02

- [deps] update `eslint-config-airbnb` to `v15.1.0`
- [deps] update `eslint-config-yoctol-base` to `v0.13.0`
- [deps] update `eslint-plugin-jsx-a11y` to `v5.0.3`
- [deps] update `eslint-plugin-react` to `v7.1.0`
- [deps] Allow eslint v4

## 0.13.0 / 2017-05-15

- [deps] update `eslint-config-airbnb` to `v15.0.0`
- [deps] update `eslint-config-yoctol-base` to `v0.12.5`
- [deps] update `eslint-plugin-jsx-a11y` to `v5.0.1`
- [deps] update `eslint-plugin-react` to `v7.0.1`

## 0.12.2 / 2017-04-05

- [deps] update `eslint` to `v3.19.0`
- [deps] update `eslint-config-yoctol-base` to `v0.12.4`

## 0.12.1 / 2017-03-27

- [deps] update `eslint`, `eslint-plugin-react`, `eslint-config-airbnb`, `eslint-config-yoctol-base`, `eslint-plugin-jsx-a11y`

## 0.12.0 / 2017-01-09

- [breaking][deps] update `eslint`, `eslint-plugin-import`, `eslint-plugin-react`, `eslint-config-airbnb`, `eslint-config-yoctol-base`
- [breaking][deps] update `eslint-plugin-jsx-a11y` to v3
- [docs] add note about `install-peerdeps`
- [docs] Updated instructions to support non-bash users

## 0.11.0 / 2016-12-09

- [deps] update `eslint-config-yoctol-base` to `v0.11.0`

## 0.10.0 / 2016-11-08

- [deps] update `eslint` to `v3.9.1`
- [deps] update `eslint-plugin-import` to `v2.1.0`
- [deps] update `eslint-plugin-jsx-a11y` to `v2.2.3`
- [deps] update `eslint-plugin-react` to `v6.6.0`
- [deps] update `eslint-config-airbnb` to `v13.0.0`
- [deps] update `eslint-config-yoctol-base` to `v0.10.0`
- [docs] ensure latest version of config is installed

## 0.9.0 / 2016-09-29

- [deps] update `eslint` to v3.6.0
- [deps] update `eslint-plugin-import` to v1.16.0
- [deps] update `eslint-plugin-jsx-a11y` to v2.2.2
- [deps] update `eslint-plugin-react` to v6.3.0
- [deps] update `eslint-config-airbnb` to v12.0.0
- [deps] update `eslint-config-yoctol-base` to v0.8.0
- [breaking] enable `react/sort-prop-types`

## 0.8.1 / 2016-09-19

- [deps] update `eslint` to v3.5.0
- [deps] update `eslint-config-airbnb` to v11.1.0
- [deps] update `eslint-config-yoctol-base` to v0.7.0

## 0.8.0 / 2016-09-08

- [deps] update `eslint` to v3.4.0
- [deps] update `eslint-plugin-import` to v1.14.0
- [deps] update `eslint-plugin-jsx-a11y` to v2.2.1
- [deps] update `eslint-plugin-react` to v6.2.0
- [deps] update `eslint-config-airbnb` to v11.0.0
- [deps] update `eslint-config-yoctol-base` to v0.6.0

## 0.7.0 / 2016-09-01

- [deps] update `eslint` to v3.3.1

rules/react.js
module.exports = {
  plugins: ['react', 'react-hooks'],
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
  },

  // View link below for react rules documentation
  // https://github.com/yannickcr/eslint-plugin-react#list-of-supported-rules
  rules: {
    // Enforce consistent usage of destructuring assignment of props, state, and context
    // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/destructuring-assignment.md
    // Proposal - https://github.com/Yoctol/eslint-config-yoctol/issues/94
    'react/destructuring-assignment': 'off',

    // Enforce propTypes declarations alphabetical sorting
    // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/sort-prop-types.md
    // Proposal - https://github.com/Yoctol/eslint-config-yoctol/issues/30
    'react/sort-prop-types': [
      'error',
      {
        callbacksLast: true,
        ignoreCase: false,
        requiredFirst: false,
      },
    ],

    // Prevent usage of deprecated methods
    // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-deprecated.md
    'react/no-deprecated': ['warn'],

    // Optional stateless functions when not using lifecycle methods, setState or ref
    // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-stateless-function.md
    // Proposal - https://github.com/Yoctol/eslint-config-yoctol/issues/24
    'react/prefer-stateless-function': 'off',

    // only .js files may have JSX
    // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-filename-extension.md
    // Proposal - https://github.com/Yoctol/eslint-config-yoctol/issues/21
    'react/jsx-filename-extension': ['error', { extensions: ['.js'] }],

    // https://github.com/facebook/react/tree/master/packages/eslint-plugin-react-hooks
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'error',

    // turn off react rules that will conflict with prettier
    'react/jsx-curly-newline': 'off',
    'react/jsx-indent': 'off',
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-wrap-multilines': 'off',
  },
};
